Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift

```diff
@@ -9,201 +9,1069 @@
+        }
 
-    public mutating func next() async rethrows -> Element? {
-      return try await state.next()?.value
+        /// The state machine's current state.
+        private var state: State
+
+        /// Initializes a new `StateMachine`.
+        init(base1: Base1, base2: Base2) {
+            state = .initial(
+                base1: base1,
+                base2: base2
+            )
+        }
+
+        mutating func sequenceDeinitialized() {
+            switch state {
+            case .initial:
+                // The references to the sequence were dropped before any iterator was ever created
+                state = .finished
+
+            case .merging, .upstreamThrew:
+                // An iterator was created and we deinited the sequence.
+                // This is an expected pattern and we just continue on normal.
+                // Importantly since we are a unicast sequence no more iterators can be created
+                break
+
+            case .finished:
+                // We are already finished so there is nothing left to clean up.
+                // This is just the references dropping afterwards.
+                break
+
+            case .modifying:
+                preconditionFailure("Invalid state")
+            }
+        }
+
+        /// Actions returned by `iteratorInitialized()`.
+        enum IteratorInitializedAction {
+            /// Indicates that a new `Task` should be created that consumed the sequences.
+            case startTask(Base1, Base2)
+        }
+
+        mutating func iteratorInitialized() -> IteratorInitializedAction {
+            switch state {
+            case let .initial(base1, base2):
+                // This is the first iterator being created and we need to create our `Task`
+                // that is consuming the upstream sequences.
+                return .startTask(base1, base2)
+
+            case .merging, .upstreamThrew, .finished:
+                fatalError("merge allows only a single AsyncIterator to be created")
+
+            case .modifying:
+                preconditionFailure("Invalid state")
+            }
+        }
+
+        /// Actions returned by `iteratorDeinitialized()`.
+        enum IteratorDeinitializedAction {
+            /// Indicates that the `Task` needs to be cancelled and
+            /// all upstream continuations need to be resumed with a `CancellationError`.
+            case cancelTaskAndUpstreamContinuations(
+                task: Task<Void, Never>,
+                upstreamContinuations: [CheckedContinuation<Void, Error>]
+            )
+            /// Indicates that nothing should be done.
+            case none
+        }
+
+        mutating func iteratorDeinitialized() -> IteratorDeinitializedAction {
+            switch state {
+            case .initial:
+                // An iterator needs to be initialized before it can be deinitialized.
+                preconditionFailure("Internal inconsistency")
```

Comment:
the preconditions in here make my spidey senses tingle.... any fatal error will become a bug (in my experience). is it possible to somehow refactor these to not be?

Reply:
IMO, we should keep these preconditions. The state that I am preconditioning in should never been hit and I haven't been able to force the state machine into these states with my tests. If we ever hit one of those, I consider this a good thing because the assumptions I have made are wrong.

The good thing is this won't become a subtle bug but rather a crash. We have been using this in NIO for a long time and it helped us flush out bugs in all our state machines.